Experiment: switch linux-arm64v8 from cross-compilation to native build

It looks like Debian 11 (Bullseye) has upgraded from glibc v2.29 to v2.31 as part of its promotion to "stable".

https://packages.debian.org/bullseye/libc6

glibc v2.29 was already a bit too "modern" and I expect v2.31 will cause even more problems.

Given we now have access to an ARM64 CPU via CircleCI, this future possible experiment is to try using a native build for the linux-arm64v8 platform, hopefully one that can still be run via QEMU on x64 if necessary.

I took a quick look at this via commit https://github.com/lovell/sharp-libvips/commit/7d56723c93b407f339ae97d9dd59640d63cb1b56 but librsvg fails to build due to our use of `-Zbuild-std` on ARM64.
```
cargo --locked build   --release --lib \
&& cd /deps/svg && /bin/sh ./libtool --silent --tag=CC   --mode=link gcc  -march=armv8-a -Os -fPIC -D_GLIBCXX_USE_CXX11_ABI=1 -fno-asynchronous-unwind-tables -ffunction-sections -fdata-sections  -L/target/lib -Wl,--gc-sections -Wl,-rpath='$ORIGIN/' -o librsvg_c_api.la _rsvg_dummy.lo && mv /deps/svg/target/release/liblibrsvg.a .libs/librsvg_c_api.a
error: -Zbuild-std requires --target
make[3]: *** [librsvg_c_api.la] Error 101
```
...which suggests we might have run into https://github.com/rust-lang/cargo/issues/9451

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,0 +1,26 @@
+version: 2.1
+orbs:
+  github-cli: circleci/github-cli@1.0
+
+workflows:
+  build:
+    jobs:
+      - linux-arm64v8:
+          filters:
+            tags:
+              only: /^v.*/
+
+jobs:
+  linux-arm64v8:
+    resource_class: arm.medium
+    machine:
+      image: ubuntu-2004:202101-01
+    steps:
+      - checkout
+      - github-cli/setup
+      - run: ./build.sh $(cat LIBVIPS_VERSION) linux-arm64v8
+      - run: ./integrity.sh
+      - when:
+          condition: <<pipeline.git.tag>>
+          steps:
+            - run: gh release upload --repo lovell/sharp-libvips $CIRCLE_TAG *.tar.gz *.tar.br *.integrity

--- a/.github/workflows/build-release.yml
+++ b/.github/workflows/build-release.yml
@@ -27,7 +27,6 @@ jobs:
           - 'linux-x64'
           - 'linux-armv6'
           - 'linux-armv7'
-          - 'linux-arm64v8'
           - 'linuxmusl-x64'
           - 'linuxmusl-arm64v8'
           - 'win32-ia32'

--- a/build/lin.sh
+++ b/build/lin.sh
@@ -455,6 +455,10 @@ sed -i'.bak' "s/^\(Requires:.*\)/\1 cairo-gobject pangocairo/" librsvg.pc.in
 sed -i'.bak' "s/, \"rlib\"//" Cargo.toml
 # Skip executables
 sed -i'.bak' "/SCRIPTS = /d" Makefile.in
+# Use target/CARGO_BUILD_TARGET/release instead of target/release when set
+if [[ $CARGO_BUILD_TARGET ]]; then
+  sed -i'.bak' "s/@RUST_TARGET_SUBDIR@/$CARGO_BUILD_TARGET\/@RUST_TARGET_SUBDIR@/" Makefile.in
+fi
 ./configure --host=${CHOST} --prefix=${TARGET} --enable-static --disable-shared --disable-dependency-tracking \
   --disable-introspection --disable-tools --disable-pixbuf-loader --disable-nls --without-libiconv-prefix --without-libintl-prefix \
   ${DARWIN:+--disable-Bsymbolic}

--- a/linux-arm64v8/Dockerfile
+++ b/linux-arm64v8/Dockerfile
@@ -1,40 +1,33 @@
-FROM debian:bullseye
+FROM arm64v8/centos:7
 LABEL maintainer="Lovell Fuller <npm@lovell.info>"
 
-# Create Debian-based container suitable for cross-compiling Linux ARM64v8-A binaries
+# Create CentOS 7 (glibc 2.17) container suitable for building Linux ARM64v8-A binaries
 
 # Path settings
 ENV \
   RUSTUP_HOME="/usr/local/rustup" \
   CARGO_HOME="/usr/local/cargo" \
-  PATH="/usr/local/cargo/bin:$PATH"
+  PATH="/usr/local/cargo/bin:/opt/rh/devtoolset-10/root/usr/bin:$PATH"
 
 # Build dependencies
 RUN \
-  apt-get update && \
-  apt-get install -y curl && \
-  dpkg --add-architecture arm64 && \
-  apt-get update && \
-  apt-get install -y \
+  yum update -y && \
+  yum install -y epel-release centos-release-scl && \
+  yum group install -y "Development Tools" && \
+  yum install -y --setopt=tsflags=nodocs \
     advancecomp \
-    autoconf \
-    autopoint \
     brotli \
-    cmake \
-    crossbuild-essential-arm64 \
-    gettext \
-    git \
-    gobject-introspection \
+    cmake3 \
+    devtoolset-10-gcc \
+    devtoolset-10-gcc-c++ \
+    glib2-devel \
+    gobject-introspection-devel \
     gperf \
-    gtk-doc-tools \
-    intltool \
+    gtk-doc \
     jq \
-    libglib2.0-dev \
-    libtool \
     nasm \
-    ninja-build \
-    python3-pip \
-    texinfo \
+    prelink \
+    python3 \
     && \
   curl https://sh.rustup.rs -sSf | sh -s -- -y \
     --no-modify-path \
@@ -43,7 +36,9 @@ RUN \
     && \
   rustup component add rust-src && \
   rustup target add aarch64-unknown-linux-gnu && \
-  pip3 install meson
+  ln -s /usr/bin/cmake3 /usr/bin/cmake && \
+  pip3 install --upgrade pip && \
+  pip3 install meson ninja
 
 # Rebuild the standard library of Rust to avoid collisions with system libraries.
 # See: https://github.com/lovell/sharp-libvips/issues/109
@@ -54,10 +49,8 @@ build-std-features = [\"panic_immediate_abort\"]" > $CARGO_HOME/config.toml
 
 # Compiler settings
 ENV \
-  PKG_CONFIG="/usr/bin/aarch64-linux-gnu-pkg-config" \
   PLATFORM="linux-arm64v8" \
-  CHOST="aarch64-linux-gnu" \
-  RUST_TARGET="aarch64-unknown-linux-gnu" \
+  CARGO_BUILD_TARGET="aarch64-unknown-linux-gnu" \
   FLAGS="-march=armv8-a" \
   MESON="--cross-file=/root/meson.ini"
 

--- a/linux-arm64v8/Toolchain.cmake
+++ b/linux-arm64v8/Toolchain.cmake
@@ -1,13 +1,3 @@
-set(CMAKE_SYSTEM_NAME Linux)
-set(CMAKE_SYSTEM_VERSION 1)
-set(CMAKE_SYSTEM_PROCESSOR aarch64)
-
-SET(CMAKE_C_COMPILER /usr/bin/aarch64-linux-gnu-gcc)
-set(CMAKE_CXX_COMPILER /usr/bin/aarch64-linux-gnu-g++)
-SET(CMAKE_AR /usr/bin/aarch64-linux-gnu-gcc-ar)
-SET(CMAKE_STRIP /usr/bin/aarch64-linux-gnu-gcc-strip)
-SET(CMAKE_RANLIB /usr/bin/aarch64-linux-gnu-gcc-ranlib)
-
 set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
 set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
 set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)

--- a/linux-arm64v8/meson.ini
+++ b/linux-arm64v8/meson.ini
@@ -1,18 +1,6 @@
-[host_machine]
-system = 'linux'
-cpu_family = 'aarch64'
-cpu = 'armv8-a'
-endian = 'little'
-
 [binaries]
-c = '/usr/bin/aarch64-linux-gnu-gcc'
-cpp = '/usr/bin/aarch64-linux-gnu-g++'
-ar = '/usr/bin/aarch64-linux-gnu-gcc-ar'
-nm = '/usr/bin/aarch64-linux-gnu-gcc-nm'
-ld = '/usr/bin/aarch64-linux-gnu-gcc-ld'
-strip = '/usr/bin/aarch64-linux-gnu-strip'
-pkgconfig = '/usr/bin/aarch64-linux-gnu-pkg-config'
-ranlib = '/usr/bin/aarch64-linux-gnu-gcc-ranlib'
+pkgconfig = '/usr/bin/pkg-config'
+strip = '/usr/bin/strip'
 
 [paths]
 libdir = 'lib'

